compiler/src/language_server/definition.re

```diff
@@ -53,38 +53,69 @@ let send_definition =
     }),
   );
 };
+type check_position =
+  | Forward
+  | Backward;
+let rec find_definition =
+        (
+          ~check_position=Forward,
+          sourcetree: Sourcetree.sourcetree,
+          position: Protocol.position,
+        ) => {
+  let results = Sourcetree.query(position, sourcetree);
 
-let process =
-    (
-      ~id: Protocol.message_id,
-      ~compiled_code: Hashtbl.t(Protocol.uri, Lsp_types.code),
-      ~documents: Hashtbl.t(Protocol.uri, string),
-      params: RequestParams.t,
-    ) => {
-  switch (Hashtbl.find_opt(compiled_code, params.text_document.uri)) {
-  | None => send_no_result(~id)
-  | Some({program, sourcetree}) =>
-    let results = Sourcetree.query(params.position, sourcetree);
-
+  let result =
     switch (results) {
     | [Value({definition}), ..._]
     | [Pattern({definition}), ..._]
     | [Type({definition}), ..._]
     | [Declaration({definition}), ..._]
     | [Module({definition}), ..._] =>
       switch (definition) {
-      | None => send_no_result(~id)
+      | None => None
       | Some(loc) =>
         let uri = Utils.filename_to_uri(loc.loc_start.pos_fname);
-
-        send_definition(
-          ~id,
-          ~range=Utils.loc_to_range(loc),
-          ~target_uri=uri,
-          Utils.loc_to_range(loc),
-        );
+        Some((loc, uri));
       }
-    | _ => send_no_result(~id)
+    | _ => None
+    };
+  switch (result) {
+  | None =>
+    if (check_position == Forward && position.character > 0) {
+      // If a user selects from left to right, their pointer ends up after the identifier
+      // this trys to check if the identifier was selected.
+      find_definition(
+        ~check_position=Backward,
+        sourcetree,
+        {line: position.line, character: position.character - 1},
+      );
```

Comment:
I wonder if this is actually just a "bug" in the sourcetree, as such the sourcetree seems to be position exclusive when it probably should be location *inclusive*. Is this a correct assessment?

Reply:
I don't think so the issue really seems to be from highlighting and where your cursor is. we do not know which way the user is highlighting left to right or right to left. as such we don't know where the cursor will be.

Reply:
The sourcetree is supposed to provide a correct node, no matter the position of the cursor. Whether the cursor is like: `|foobar` or `foobar|` or `foo|bar` or anything. And it seems that it doesn't.

Reply:
> The sourcetree is supposed to provide a correct node, no matter the position of the cursor. Whether the cursor is like: `|foobar` or `foobar|` or `foo|bar` or anything. And it seems that it doesn't.

Digging into this a bit more If we want to fix this in `sourceTree` then I think we should just add `1` to `pos_end.charater` in `loc_to_interval`. I still think this is an issue in the lsp spec. mainly because of the case of `foo|+1` with the cursor location being represented by `|` its ambigous if the user is selecting over `foo` from left to right or selecting `+` from right to left.

Reply:
@ospencer does Jake's thoughts above sound correct here?

Reply:
Jake is correct. The LSP spec doesn't account for this situation pretty much at all—if your cursor is on the edges of the value, it's not overlapping the range.

Reply:
If the LSP spec doesn't say that it's in the range, then I think we should close this PR.

Reply:
It's up to our interpretation. The editor will send the request with whatever the cursor position is. We don't consider a cursor after something as overlapping that range. That's the correct behavior, but in the case Jake is trying to address—highlighting an identifier from the beginning to the end—your cursor ends up after, and the goto fails.

We have to decide if we care. I think it's reasonable enough to try to improve the user experience because I imagine some users do this. I don't think we should change the sourcetree though, because it's correct.

Reply:
I agree about not changing the sourcetree. For context as to when this was a problem for me. I was double clicking on an identifier to select it and then clicking goto definition which I feel is common behavior.

Reply:
> We don't consider a cursor after something as overlapping that range.

If *we don't* consider the cursor in the range, then **one** of our tools shouldn't either.

Reply:
I agree with the statement and when considering the cursor position I agree with the viewpoint. The issue in this case isn't the cursor position though it is the selection range. And the fact that when you double click on an identifier to select it, it puts the cursor at the end. Which seems like reasonable user behaviour.